sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala

```diff
@@ -0,0 +1,278 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import java.util
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalog.v2.{Identifier, TableCatalog, TableChange}
+import org.apache.spark.sql.catalog.v2.expressions.{BucketTransform, FieldReference, IdentityTransform, LogicalExpressions, Transform}
+import org.apache.spark.sql.catalog.v2.utils.CatalogV2Util
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.analysis.{NoSuchNamespaceException, NoSuchTableException, TableAlreadyExistsException}
+import org.apache.spark.sql.catalyst.catalog.{BucketSpec, CatalogTable, CatalogTableType, CatalogUtils, SessionCatalog}
+import org.apache.spark.sql.execution.datasources.DataSource
+import org.apache.spark.sql.internal.SessionState
+import org.apache.spark.sql.sources.v2.{Table, TableCapability, TableProvider}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+/**
+ * A [[TableCatalog]] that translates calls to the v1 SessionCatalog.
+ */
+class V2SessionCatalog(sessionState: SessionState) extends TableCatalog {
+  def this() = {
+    this(SparkSession.active.sessionState)
+  }
+
+  private lazy val catalog: SessionCatalog = sessionState.catalog
+
+  private var _name: String = _
+
+  override def name: String = _name
+
+  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
+    this._name = name
+  }
+
+  override def listTables(namespace: Array[String]): Array[Identifier] = {
+    namespace match {
+      case Array(db) =>
+        catalog.listTables(db).map(ident => Identifier.of(Array(db), ident.table)).toArray
+      case _ =>
+        throw new NoSuchNamespaceException(namespace)
+    }
+  }
+
+  def loadTable(ident: Identifier): Table = {
+    val catalogTable = try {
+      catalog.getTableMetadata(ident.asTableIdentifier)
+    } catch {
+      case _: NoSuchTableException =>
+        throw new NoSuchTableException(ident)
+    }
+
+    catalogTable.provider match {
+      case Some(provider) =>
+        DataSource.lookupDataSource(provider, sessionState.conf).newInstance() match {
+          case tableProvider: TableProvider =>
+            val storageOptions = ident.namespace match {
+              case Array(db) =>
+                catalogTable.storage.properties +
+                    ("provider" -> provider) +
+                    ("database" -> db) +
+                    ("table" -> ident.name)
+              case Array() =>
+                catalogTable.storage.properties +
+                    ("provider" -> provider) +
+                    ("database" -> catalog.getCurrentDatabase) +
+                    ("table" -> ident.name)
+            }
+            tableProvider.getTable(new CaseInsensitiveStringMap(storageOptions.asJava))
```

Comment:
Should we pass the catalog table schema to `getTable`? 
```
            val tableOptions = new CaseInsensitiveStringMap(storageOptions.asJava)
            tableProvider.getTable(tableOptions, catalogTable.schema)
```
In such way, [SPARK-27960](https://issues.apache.org/jira/browse/SPARK-27960) is also resolved.

Reply:
No, that is not the correct solution for SPARK-27960.

The user supplied schema is used when interpreting data that has no schema, like CSV or JSON. That's why it is called the user-supplied schema.

In this case, the schema is stored in the metastore. The table provider can access it from the session metastore using the table and database options.

I agree that this is strange and isn't a good way to handle these tables. I think it would be a good idea to change the `TableProvider` interface to accept more metadata from the metastore, but we should consider those options and should not reuse the user-supplied schema for something other than what the API was intended for.

Reply:
> In this case, the schema is stored in the metastore. The table provider can access it from the session metastore using the table and database options.

In such case, we call `V2SessionCatalog.loadTable` to get a table, while the table provider has to access metastore again to get schema. Seems tricky.

>  I think it would be a good idea to change the TableProvider interface to accept more metadata from the metastore, but we should consider those options and should not reuse the user-supplied schema for something other than what the API was intended for.

I admit that reusing the user-specified schema sounds a bit ugly. But it is a simple solution. There is similar usage  in `FindDataSourceTable`/`CreateDataSourceTableCommand` when constructing `DataSource`. 
I believe that there could be a better solution than this. (Maybe we can start another thread for this.)

Reply:
@gengliangwang, we can talk about how to fix `TableProvider` later. It should not block this commit.

Reply:
OK, we can resolve this later.